### PR TITLE
Fix #11300

### LINKED ISSUE #11300: 柱状图dataZoom和markPoint同时使用,dataZoom放大到只显示某一个柱的时候,所有的markPoint同时出现在这个柱上
### Version
4.3.0

### Reproduction link
[https://jsfiddle.net/v8r3dx02/2/](https://jsfiddle.net/v8r3dx02/2/)

### Steps to reproduce
放大dataZoom到视图只显示一个柱

### What is expected?
markPoint正常显示

### What is actually happening?
所有的markPoint全部出现

### PR DESCRIPTION
**Pull request by susiwen8**
For category type, one data show on window just check if scale contains it's value

Close #11300

### CODE REVIEW

**Comment by susiwen8** [comment]
I got problem to install puppeteer, so this PR hasn't test with test cases.

File: src/coord/cartesian/Cartesian2D.js [hunk 1]
```diff
@@ -81,7 +81,15 @@ Cartesian2D.prototype = {
      * @return {boolean}
      */
     containData: function (data) {
-        return this.getAxis('x').containData(data[0])
+        var axisX = this.getAxis('x');
+
+        // For aixs type is category,
+        // only one data shows on viewport just check if it xAxis' value equal to scale
+        if (axisX.type === 'category' && axisX.scale.count() === 1) {
+            return data[0] === axisX.scale._extent[0];
```
Comment (Ovilia) [inline src/coord/cartesian/Cartesian2D.js]: `_extent` is a private variable. You should probably call `axisX.scale.getExtent()[0]` to get it.

File: src/coord/cartesian/Cartesian2D.js [hunk 2]
```diff
@@ -81,7 +81,15 @@ Cartesian2D.prototype = {
...
+        }
+
+        return axisX.containData(data[0])
```
Comment (Ovilia) [inline src/coord/cartesian/Cartesian2D.js]: I'm not sure if you should fix the implementation of `axisX.containData` in the case of one data item.

**Comment by pissang** [comment]
Hi @susiwen8 .  

I think the main cause is from https://github.com/apache/incubator-echarts/blob/master/src/coord/Axis.js#L88

Current solution of `containData` will convert the data to coordinates, then verify if coordinates is between the axis extent. Some information will be lost during this conversion. Like coord will be clamped as you metioned in https://github.com/apache/incubator-echarts/issues/11300#issuecomment-536280273

From my perspective, a better solution is using `Scale#contain` 

```js
containData: function (data) {
  return this.scale.contain(data);
}
```

BTW: You can try using `cnpm` to install puppeteer. It will use Taobao Mirror to download the executable.

**Comment by susiwen8** [comment]
@pissang Hi, I have tried `this.scale.contain(data)`, for x value is fine,  but not for y value. Because yAxis' extent only consider the data not the position of markpoint.

**Comment by pissang** [comment]
Hi @susiwen8 sorry for the late reply.

> Because yAxis' extent only consider the data not the position of markpoint
I didn't understand this. Did you mean `value` type axis?
It will be great if you can give an example to explain it. I tried using `this.scale.contain(data)` and it works well on the example in issue.

**Comment by susiwen8** [comment]
Screen Shot 2019-11-06 at 13 28 47
Screen Shot 2019-11-06 at 13 28 57
Screen Shot 2019-11-06 at 13 29 07
Screen Shot 2019-11-06 at 13 29 12
@pissang Because the yAxis of markpoint for Aug and Dec are way higher than value, they aren't fit in Y axis if we use scale. Scale only consider data not markpoint.

**Comment by pissang** [comment]
Understand. I think it's another issue: If markPoint should be clipped if it exceed the axis. 

But in this issue. Using `this.scale.contain(data)` would make more sense for the method `containData`.

**Pull request merged**